deku/index.d.ts

```diff
@@ -0,0 +1,137 @@
+// Project: https://github.com/anthonyshort/deku
+// Definitions by: Sho Fuji <https://github.com/pocka/>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export = deku;
+export as namespace deku;
+
+declare namespace deku {
+
+	interface VirtualElement {
+		type: string;
+	}
+
+	/**
+	 * Create a DOM renderer using a container element.
+	 * Everything will be rendered inside of that container.
+	 * Returns a function that accepts new state that can replace what is currently rendered.
+	 */
+	function createApp(el: HTMLElement): Render;
+	function createApp(el: HTMLElement, dispatch: Function): Render;
+
+	namespace dom {
+		/**
+		 * Create a real DOM element from a virtual element, recursively looping down.
+		 * When it finds custom elements it will render them, cache them, and keep going,
+		 * so they are treated like any other native element.
+		 */
+		function create<C>(vnode: VirtualElement, path: string, dispatch: Function, context: C): HTMLElement;
+
+		/**
+		 * Modify a DOM element given an array of actions.
+		 */
+		function update<C, A>(dispatch: Function, context: C): (DOMElement: HTMLElement, action: A) => HTMLElement;
+	}
+
+	namespace string {
+		/**
+		 * Render a virtual element to a string. You can pass in an option state context object that will be given to all components.
+		 */
+		function render(vnode: VirtualElement): string;
+		function render<C>(vnode: VirtualElement, context: C): string;
+	}
+
+	/**
+	 * This function lets us create virtual nodes using a simple syntax.
+	 * It is compatible with JSX transforms so you can use JSX to write nodes that will compile to this function.
+	 */
+	function element(type: string): VirtualElement;
+	function element<A>(type: string, attributes: A): VirtualElement;
+	function element<A>(type: string, attributes: A, ...children: any[]): VirtualElement;
+
+	function element(type: Thunk): VirtualElement;
+	function element<A>(type: Thunk, attributes: A): VirtualElement;
+	function element<A>(type: Thunk, attributes: A, ...children: any[]): VirtualElement;
+
+	var h: typeof element;
+
+	namespace diff {
+		/**
+		 * Compare two virtual nodes and return an array of changes to turn the left into the right.
+		 */
+		function diffNode(prevNode: VirtualElement, nextNode: VirtualElement): any[];
+
+		namespace Actions {
+			function setAttribute(a: string, b: any, c: any): any;
+			function removeAttribute(a: string, b: any): any;
+			function insertChild(a: any, b: number, c: string): any;
+			function removeChild(a: number): any;
+			function updateChild(a: number, b: any[]): any;
+			function updateChildren(a: any[]): any;
+			function insertBefore(a: number): any;
+			function replaceNode(a: any, b: any, c: string): any;
+			function removeNode(a: any): any;
+			function sameNode(): any;
+			function updateThunk(a: any, b: any, c: string): any;
+
+			// Since "case" is a reserved word, we cannot use "deku.Actions.case" in TypeScript.
```

Comment:
`Action` is generated by [union-type](https://github.com/paldepind/union-type) which appears to generate a class-like, so you can use:
```ts
class Actions {
	static setAttribute(a: string, b: any, c: any): Actions;
	static removeAttribute(a: string, b: any): Actions;
	static insertChild(a: any, b: number, c: string): Actions;
	static removeChild(a: number): Actions;
	static updateChild(a: number, b: any[]): Actions;
	static updateChildren(a: any[]): Actions;
	static insertBefore(a: number): Actions;
	static replaceNode(a: any, b: any, c: string): Actions;
	static removeNode(a: any): Actions;
	static sameNode(): Actions;
	static updateThunk(a: any, b: any, c: string): Actions;
	static case(pat: any, action: any): void;
}
```
It's also possible to statically type `case`:

```ts
interface CasePattern<T> {
	setAttribute(a: string, b: any, c: any): T;
	removeAttribute(...): T;
	...
}

class Action {
	...
	static case<T>(pat: CasePattern<T>, action: Action): T
}
```

...but you might not consider that worth the verbosity.